tests/integration/test_tls.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+
+import pytest
+from juju.application import Application
+from pytest_operator.plugin import OpsTest
+
+from literals import INTERNAL_USER, PEER_RELATION
+
+from .helpers import (
+    APP_NAME,
+    get_cluster_endpoints,
+    get_cluster_members,
+    get_juju_leader_unit_name,
+    get_key,
+    get_secret_by_label,
+    put_key,
+)
+
+logger = logging.getLogger(__name__)
+
+TLS_NAME = "self-signed-certificates"
+NUM_UNITS = 3
+TEST_KEY = "test_key"
+TEST_VALUE = "42"
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy_with_tls(ops_test: OpsTest) -> None:
+    """Build the charm-under-test and deploy it with three units.
+
+    The initial cluster should be formed and accessible.
+    """
+    assert ops_test.model is not None
+    # Deploy the TLS charm
+    tls_config = {"ca-common-name": "etcd"}
+    await ops_test.model.deploy(TLS_NAME, channel="edge", config=tls_config)
+    # Build and deploy charm from local source folder
+    etcd_charm = await ops_test.build_charm(".")
+    model = ops_test.model_full_name
+    assert model is not None
+    # Deploy the charm and wait for active/idle status
+    await ops_test.model.deploy(etcd_charm, num_units=NUM_UNITS)
+
+    # enable TLS and check if the cluster is still accessible
+    await ops_test.model.integrate(f"{APP_NAME}:peer-certificates", TLS_NAME)
+    await ops_test.model.integrate(f"{APP_NAME}:client-certificates", TLS_NAME)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
```

Comment:
In another PR - Please move away from `wait_for_idle` in favor of [`wait_until`](https://github.com/canonical/opensearch-operator/blob/2/edge/tests/integration/helpers_deployments.py#L348). This will give us more fine grained control over the app + unit statuses.